src/main/java/commission/dao/SaleDaoImpl.java

```diff
@@ -6,43 +6,42 @@
 import org.springframework.jdbc.core.JdbcTemplate;
 import org.springframework.stereotype.Repository;
 
+import javax.sql.DataSource;
 import java.util.List;
 import java.util.Optional;
 
 @Repository
 public class SaleDaoImpl implements SaleDao{
 
-    JdbcTemplate jdbcTemplate;
+    private final JdbcTemplate jdbcTemplate;
+    private final SalesRowMapper rowMapper = new SalesRowMapper();
 
     @Autowired
-    public SaleDaoImpl(JdbcTemplate jdbcTemplate) {
-        this.jdbcTemplate = jdbcTemplate;
+    public SaleDaoImpl(DataSource dataSource) {
+        this.jdbcTemplate = new JdbcTemplate(dataSource);
     }
 
+
     @Override
     public List<Sale> findAll() {
         var query = "SELECT * FROM sales";
-        return jdbcTemplate.query(query, new SalesRowMapper());
+        return jdbcTemplate.query(query, rowMapper);
     }
 
 
     @Override
-    public void insertSale(Sale sale) {
+    public void insertSale(Sale sale, double commission) {
         var query = "INSERT INTO sales(price, percentage, commission) VALUES (?,?,?)";
-        jdbcTemplate.update(query, sale.price(), sale.percentage(), commissionPerSale(sale));
+        jdbcTemplate.update(query, sale.price(), sale.percentage(), commission);
     }
 
     @Override
     public Optional<Sale> findSaleById(long id) {
         var query = "SELECT * FROM sales WHERE id=?";
-        return Optional.ofNullable(jdbcTemplate.queryForObject(query, new SalesRowMapper(), id));
+        return Optional.ofNullable(jdbcTemplate.queryForObject(query, rowMapper, id));
```

Comment:
You can either remove `Optional` from the return type or you have to wrap this in a `try / catch` for the `EmptyResultDataAccessException`. The `queryForObect` method does not return `null`.

You can use `Optional.empty()` if you want to go the `try/catch` route. I would suggest that, because otherwise it forces the service layer to handle the exception where as with `Optional` you can check that it exists or not

Reply:
Ok got it. I did play around with it now too, just to get it clear in my head. Hope I went the right way

Reply:
Yup it's correct, just a comment on the type of exception being caught